packages/fulfillment/src/services/fulfillment-module-service.ts

```diff
@@ -671,7 +746,137 @@ export default class FulfillmentModuleService<
   ): Promise<
     FulfillmentTypes.ShippingOptionDTO[] | FulfillmentTypes.ShippingOptionDTO
   > {
-    return []
+    const updatedShippingOptions = await this.updateShippingOptions_(
+      data,
+      sharedContext
+    )
+
+    return await this.baseRepository_.serialize<
+      FulfillmentTypes.ShippingOptionDTO | FulfillmentTypes.ShippingOptionDTO[]
+    >(updatedShippingOptions, {
+      populate: true,
+    })
+  }
+
+  @InjectTransactionManager("baseRepository_")
+  async updateShippingOptions_(
+    data:
+      | FulfillmentTypes.UpdateShippingOptionDTO[]
+      | FulfillmentTypes.UpdateShippingOptionDTO,
+    @MedusaContext() sharedContext: Context = {}
+  ): Promise<TShippingOptionEntity | TShippingOptionEntity[]> {
+    const data_ = Array.isArray(data) ? data : [data]
+
+    if (!data_.length) {
+      return []
+    }
+
+    const shippingOptionIds = data_.map((s) => s.id)
+    if (!shippingOptionIds.length) {
+      return []
+    }
+
+    const shippingOptions = await this.shippingOptionService_.list(
+      {
+        id: shippingOptionIds,
+      },
+      {
+        relations: ["rules"],
+      },
+      sharedContext
+    )
+
+    const shippingOptionSet = new Set(shippingOptions.map((s) => s.id))
+    const expectedShippingOptionSet = new Set(data_.map((s) => s.id))
+    const missingShippingOptionIds = getSetDifference(
+      expectedShippingOptionSet,
+      shippingOptionSet
+    )
+
+    if (missingShippingOptionIds.size) {
+      throw new MedusaError(
+        MedusaError.Types.NOT_FOUND,
+        `The following shipping options do not exist: ${Array.from(
+          missingShippingOptionIds
+        ).join(", ")}`
+      )
+    }
+
+    const ruleIdsToDelete: string[] = []
+    data_.forEach((shippingOption) => {
```

Comment:
This piece of code is a great candidate for a refactor. Let's try to break it down into helper functions, as it's really difficult to follow the logic as it is.